packages/fxa-payments-server/src/components/fields.tsx

```diff
@@ -0,0 +1,236 @@
+  tooltip = true,
+  required = false,
+  label,
+  className='input-row',
+  children,
+}: FieldHOCProps) => {
+  const { validator } = useContext(FormContext) as FormContextValue;
+
+  // Disabling lint rule because registering a field changes the validator
+  // instance, so including validator in list of deps causes an infinite loop
+  /* eslint-disable react-hooks/exhaustive-deps */
+  useEffect(
+    () => validator.registerField({ name, initialValue, required, fieldType }),
+    [ name, required, fieldType ]
+  );
+  /* eslint-enable react-hooks/exhaustive-dep */
+
+  return (
+    <div className={className}>
+      <label>
+        {label && <span className="label-text">{label}</span>}
+        {children}
+        {tooltip && tooltipParentRef && validator.getError(name) &&
+          <Tooltip parentRef={tooltipParentRef}>{validator.getError(name)}</Tooltip>}
+      </label>
+    </div>
+  );
+};
+
+type InputProps =
+  & { onValidate?: OnValidateFunction }
+  & FieldProps
+  & React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>;
+
+export const Input = (props: InputProps) => {
+  const {
+    name,
+    label, 
+    initialValue, 
+    onValidate,
+    tooltip = true, 
+    required = false, 
+    className, 
+    ...childProps 
+  } = props;
+  const { validator } = useContext(FormContext) as FormContextValue;
+
+  const onChange = useCallback(
+    (ev) => {
+      const { value } = ev.target;
+      if (onValidate) {
+        const { value: newValue, error } = onValidate(value);
+        validator.updateField({ name, value: newValue, error, valid: error === null });
+      } else if (required && value !== null && ! value) {
+        validator.updateField({ name, value, error: 'This field is required' });
+      } else {
+        validator.updateField({ name, value, valid: true });
+      }
+    },
+    [ name, validator ]
+  );
+
+  const tooltipParentRef = useRef<HTMLInputElement>(null);
+
+  return (
+    <Field {...{ fieldType: 'input', tooltipParentRef, name, initialValue, tooltip, required, label, className }}>
+      <input {...{
+        ...childProps,
+        ref: tooltipParentRef,
+        name,
+        required,
+        className: classNames({ invalid: validator.isInvalid(name) }),
+        value: validator.getValue(name, ''),
+        onChange: onChange,
+        onBlur: onChange,
+      }} />
+    </Field>
+  );
+};
+
+type StripeElementProps = FieldProps &
+  { component: any } &
+  ReactStripeElements.ElementProps;
+
+export const StripeElement = (props: StripeElementProps) => {
+  const {
+    component: StripeElementComponent,
+    name, tooltip = true,
+    required = false,
+    label,
+    className,
+    ...childProps
+  } = props;
+  const { validator } = useContext(FormContext) as FormContextValue;
+
+  const onChange = useCallback(
+    (value: stripe.elements.ElementChangeResponse) => {
+      if (value !== null) {
+        if (value.error && value.error.message) {
```

Comment:
Oh yeah, and this is the weird thing about the Stripe elements: They're self-validating and never reveal their contents to us. So, we either see that there was an error, or that the field is complete & valid. But, our code never sees credit card numbers, exp. dates, CVC, etc. But, we can at least see enough to know whether to flag it as invalid and display a tooltip.